snippets/varia/heatscatter_example/heatscatter_example.Rmd

```diff
@@ -0,0 +1,641 @@
+    ggtitle('Shuffeled data scatterplot\nRuntime: ca 3 s')
+
+system.time({print(p)})
+
+```
+
+## 2D bining aproach
+
+My go-to approach for these kind plots is 2D binning the data.
+The image is divided into a grid of bins x bins squares and the number of objects falling into the squares are counted.
+
+Advantages:
+- Fast and scalable:
+    - Allows faster plotting iterations
+    - Doesnt plot individual overlapping points
+    - Calculating counts is a linear function of the number of points. Thus this scales to millions (or 100 millions) of points.
+- Single, separated points still visible
+- No additional dependencies
+
+Disadvantages:
+- Hard to identify individual points interactively
+- Manual choice of bins -> I recommend to choose so many bins that individual points just visible at screen resolution
+
+Neutral
+- Hard bining - no constraints on smootheness: This reports exactly how many points fall into each bin, no smoothness is assumed. Approaches that assume smoothness might be advantagous in cases
+
+Comment: solutions like datashader pretty much solve all these issues, by choose the binsize dynamically. In datashader bins are automatically choosen according to output resolution, eg allowing to zoom in, until all points are individually visible. But datashader only exists for R - as far as I know.
+
+Further this works as well with the geom_hexbin - I just happen to like squares better :)
+
+Additionally one can also plot the bins as points, to get a more 'scatterplot' like look.
+
+
+```{r}
+p <- ggplot(dists_shuffled, aes(x=x, y=y))+
+    geom_bin2d(bins=300)+
+    ggtitle('Bin2D\nDefault colors\nRuntime: ca 0.6 s')
+
+system.time({print(p)})
+
+```
+
+Also choosing a good colormap and transforming counts is key.
+
+```{r}
+p <- ggplot(dists_shuffled, aes(x=x, y=y))+
+    geom_bin2d(bins=300)+
+    scale_fill_viridis(trans = "log10", option='inferno')+
+    ggtitle('Bin2D\nOptimized colors\nRuntime: ca 0.7 s')
+
+system.time({print(p)})
+
+
+```
+
+It is also possible to 'fake' individual points using this strategy.
+This has still the advantage of avoiding overplotting:
+
+
+```{r}
+p <- ggplot(dists_shuffled, aes(x=x, y=y))+
+    stat_bin2d(aes(color=..count..), bins=300, geom='point', size=1, fill=1)+
+    scale_color_viridis(trans = "log10", option='inferno')+
+    ggtitle('Bin2D Points\nFake Points\nRuntime: ca 0.7 s')
+
+system.time({print(p)})
+
+```
+
+```{r}
+
+```
+
+This visualization is now really almost equivalent to the `geom_pointdensity` - but way faster
+and better scaling.
+
+Bins can be used to make the integration coarser or finer.
+
+
+```{r}
+p <- ggplot(dists_shuffled, aes(x=x, y=y))+
+    stat_bin2d(aes(color=..count..), bins=100, geom='point', size=1.4, fill=1)+
+    scale_color_viridis(trans = "log10", option='inferno')+
+    ggtitle('Bin2D Points\nCoarser Fake Points\nRuntime: ca 0.5 s')
+
+system.time({print(p)})
+
+
+```
+
+Whenever bin2d is uese, also geom_hexbin can be used:
```

Comment:
used

Reply:
Spelling things correctly is really not my forte - thanks will fix :)